backpack/core/derivatives/lstm.py

```diff
@@ -0,0 +1,358 @@
+            if t == T - 1:
+                C_prod[:, t] = einsum(
+                    "vnh, nh, nh -> vnh",
+                    H_prod[:, t],
+                    ifgo[t, :, H3:H4],
+                    (1 - c_tanh[t] ** 2),
+                )
+            else:
+                C_prod[:, t] = einsum(
+                    "vnh, nh, nh -> vnh",
+                    H_prod[:, t],
+                    ifgo[t, :, H3:H4],
+                    (1 - c_tanh[t] ** 2),
+                ) + einsum(
+                    "vnh, nh -> vnh",
+                    C_prod[:, t + 1],
+                    ifgo[t + 1, :, H1:H2],
+                )
+
+            IFGO_prod[:, t, :, H3:] = einsum(
+                "vnh, nh, nh -> vnh",
+                H_prod[:, t],
+                c_tanh[t],
+                ifgo[t, :, H3:H4] * (1 - ifgo[t, :, H3:H4]),
+            )
+            IFGO_prod[:, t, :, :H] = einsum(
+                "vnh, nh, nh -> vnh",
+                C_prod[:, t],
+                ifgo[t, :, H2:H3],
+                ifgo[t, :, H0:H1] * (1 - ifgo[t, :, H0:H1]),
+            )
+            if t >= 1:
+                IFGO_prod[:, t, :, H1:H2] = einsum(
+                    "vnh, nh, nh -> vnh",
+                    C_prod[:, t],
+                    c[t - 1],
+                    ifgo[t, :, H1:H2] * (1 - ifgo[t, :, H1:H2]),
+                )
+            IFGO_prod[:, t, :, H2:H3] = einsum(
+                "vnh, nh, nh -> vnh",
+                C_prod[:, t],
+                ifgo[t, :, H0:H1],
+                1 - ifgo[t, :, H2:H3] ** 2,
+            )
+        return IFGO_prod
+
+    def _jac_mat_prod(
+        self,
+        module: LSTM,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+        sum_batch: bool = True,
+    ) -> Tensor:
+        V: int = mat.shape[0]
+        T: int = mat.shape[1]
+        N: int = mat.shape[2]
+        H: int = module.hidden_size
+        H0: int = 0 * H
+        H1: int = 1 * H
+        H2: int = 2 * H
+        H3: int = 3 * H
+        H4: int = 4 * H
+
+        ifgo, c, c_tanh, h = self._forward_pass(module, mat)
+        H_prod: Tensor = zeros(V, T, N, H, device=mat.device)
+        C_prod: Tensor = zeros(V, T, N, H, device=mat.device)
+        C_tanh_prod_t: Tensor = zeros(V, N, H, device=mat.device)
+        IFGO_prod_t: Tensor = zeros(V, N, 4 * H, device=mat.device)
+        for t in range(T):
+            # product until nodes ifgo
+            if t == 0:
+                IFGO_prod_t[:] = einsum(
+                    "hi, vni -> vnh",
+                    module.weight_ih_l0,
+                    mat[:, t],
+                )
+            else:
+                IFGO_prod_t[:] = einsum(
+                    "hi, vni -> vnh",
+                    module.weight_ih_l0,
+                    mat[:, t],
+                ) + einsum(
+                    "hg, vng -> vnh",
+                    module.weight_hh_l0,
+                    H_prod[:, t - 1],
+                )
```

Comment:
Try a similar refactoring to above that reduces duplication.

Reply:
Yes, except in-place (see discussion above).